man/pVst.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "PVST" "1" "" "pVst (vcflib)" "pVst (VCF statistics)"
.hy
.SH NAME
.PP
\f[B]pVst\f[R]
.SH SYNOPSIS
.PP
\f[B]pVst\f[R] \[en]target 0,1,2,3,4,5,6,7 \[en]background
11,12,13,16,17,19,22 \[en]file my.vcf \[en]type CN
.SH DESCRIPTION
.PP
\f[B]pVst\f[R] calculates vst, a measure of CNV stratification.
.SH OPTIONS
.IP
.nf
\f[C]




The statistic Vst is used to test the difference in copy numbers at
each SV between two groups: Vst = (Vt-Vs)/Vt, where Vt is the overall
variance of copy number and Vs the average variance within
populations.

Output : 4 columns :     
     1. seqid            
     2. position         
     3. end              
     3. vst              
     4. probability      

required: t,target     -- argument: a zero based comma separated list of target individuals corresponding to VCF columns       
required: b,background -- argument: a zero based comma separated list of background individuals corresponding to VCF columns   
required: f,file       -- argument: a properly formatted VCF.                                                                  
required: y,type       -- argument: the genotype field with the copy number: e.g. CN|CNF                           
optional: r,region     -- argument: a tabix compliant genomic range : seqid or seqid:start-end                                 
optional: x,cpu        -- argument: number of CPUs [1] 
optional: n,per        -- argument: number of permutations [1000] 

Type: statistics


\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.
**vcfwave** with its companion tool **vcfcreatemulti** can take an existing VCF file that contains multiple complex overlapping and even nested alleles and, unlike Humpty Dumpty, take them apart and put them together again.
Thereby, hopefully, creating sane VCF output that is useful for analysis and getting rid of false positives.

We created these tools by including the state-of-the-art [biWFA](https://github.com/smarco/WFA2-lib) wavefront aligner.

src/pVst.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

#include "Variant.h"
#include "var.hpp"
#include "index.hpp"
#include "stats.hpp"
#include "gpatInfo.hpp"

#include <string>
#include <iostream>
#include <cmath>
#include <ctime>
#include <getopt.h>
#include <algorithm>
#include <random>
#include <sstream>


#if defined HAS_OPENMP
#include <omp.h>
#endif

using namespace std;
using namespace vcflib;

struct copyNcounts{
  std::string       seqid       ;
  std::string       type        ;
  std::string       targetV     ;
  std::string       backgroundV ;
  std::string       callers     ;

  std::stringstream results ;
  long int          pos     ;
  long int          end     ;
  std::vector<double> target, background, total;
};

int nper = 1000;

void printHelp(void){
  cerr << endl << endl;
  cerr << "INFO: help" << endl;
  cerr << "INFO: description:" << endl;
  cerr << "     pVst calculates vst, a measure of CNV stratification." << endl << endl;
     cerr << R"(

The statistic Vst is used to test the difference in copy numbers at
each SV between two groups: Vst = (Vt-Vs)/Vt, where Vt is the overall
variance of copy number and Vs the average variance within
populations.

)";

  cerr << "Output : 4 columns :     "    << endl;
  cerr << "     1. seqid            "    << endl;
  cerr << "     2. position         "    << endl;
  cerr << "     3. end              "    << endl;
  cerr << "     3. vst              "    << endl;
  cerr << "     4. probability      "    << endl  << endl;

  cerr << "INFO: usage:  pVst --target 0,1,2,3,4,5,6,7 --background 11,12,13,16,17,19,22 --file my.vcf --type CN" << endl;
  cerr << endl;
  cerr << "INFO: required: t,target     -- argument: a zero based comma separated list of target individuals corresponding to VCF columns       "  << endl;
  cerr << "INFO: required: b,background -- argument: a zero based comma separated list of background individuals corresponding to VCF columns   "  << endl;
  cerr << "INFO: required: f,file       -- argument: a properly formatted VCF.                                                                  "  << endl;
  cerr << "INFO: required: y,type       -- argument: the genotype field with the copy number: e.g. CN|CNF                           "  << endl;
  cerr << "INFO: optional: r,region     -- argument: a tabix compliant genomic range : seqid or seqid:start-end                                 "  << endl;
  cerr << "INFO: optional: x,cpu        -- argument: number of CPUs [1] " << endl ;
  cerr << "INFO: optional: n,per        -- argument: number of permutations [1000] " << endl;

  cerr << endl << "Type: statistics" << endl << endl;
  cerr << endl;

  printVersion() ;
}

double bound(double v){
  if(v <= 0.00001){
    return  0.00001;
  }
  if(v >= 0.99999){
    return 0.99999;
  }
  return v;
}
//------------------------------- SUBROUTINE --------------------------------
/*
 Function input  : vector of doubles

 Function does   : calculates the var

 Function returns: double

*/

inline double var(vector<double> & d, double mu){
  double variance = 0;

  for(const auto& it : d){
    variance += pow(it - mu,2);
  }

  return variance / (d.size() - 1);
}

//------------------------------- SUBROUTINE --------------------------------
/*
 Function input  : vector of ints

 Function does   : calculates the mean

 Function returns: double

*/


double vst(copyNcounts * d){

  double muA = mean(d->total     );
  double vA = var(d->total, muA     );

  if(vA <= 0){
    return 0;
  }

  double muT = mean(d->target    );
  double muB = mean(d->background);

  double vT = var(d->target, muT    );
  double vB = var(d->background, muB);

  double cT = double(d->target.size())     / double(d->total.size());
  double cB = double(d->background.size()) / double(d->total.size());

  double ans =  (vA - ( (cT * vT) + (cB * vB) )) / vA;

  //purge negative
  if(ans < 0){
    ans = 0;
  }

  return ans;

}

void calc(copyNcounts * d){

  double v = vst(d);

  int nsuc   = 0;
  int trials = 0;

  for(uint i = 0; i < nper; i++){
    if (nsuc > 1){
      break;
    }

    trials += 1;

    static std::mt19937 g{std::random_device{}()};

    std::shuffle(d->total.begin(), d->total.end(), g);

    int tsize = d->target.size();

    d->target.clear();
    d->background.clear();

    int counter = 0;

  	for(const auto t : d->total){
  		if(counter < tsize){
  			d->target.push_back(t);
  		}
  		else{
  			d->background.push_back(t);
  		}
  		counter+=1;
  	}

    //    std::cerr << "PER\t" << v << "\t" << vst(d) << std::endl;

    if (vst(d) >= v ){
      nsuc += 1;
    }
  }

  double p = 0;
  if(nsuc > 0){
    p = double(nsuc) / double(trials);
  }



  d->results << d->seqid
	     << "\t"
	     << d->pos
	     << "\t"
	     << d->end
	     << "\t"
	     << d->type
	     << "\t"
	     << v
	     << "\t"
	     << p ;




}



//------------------------------- SUBROUTINE --------------------------------

// gotta load the dat so that we can permute using open MP
void loadDat(copyNcounts * d,
	     std::string & type,
	     vector < map< string, vector<string> > > & target,
	     vector < map< string, vector<string> > > & background){

  for(auto& it : target){
    d->target.push_back( atof(it[type].front().c_str()) );
    d->total.push_back(  atof(it[type].front().c_str()) );

    d->targetV += it[type].front();
    d->targetV += ",";

  }
  for(auto& it : background){
    d->background.push_back( atof(it[type].front().c_str()) );
    d->total.push_back( atof(it[type].front().c_str()) );

    d->backgroundV += it[type].front();
    d->backgroundV += ",";

  }

}

int main(int argc, char** argv) {

  std::srand ( unsigned ( std::time(0) ) );

  int cpu = 1;

  // the filename

  string filename = "NA";

  // set region to scaffold

  string region = "NA";

  // using vcflib; thanks to Erik Garrison

  VariantCallFile variantFile;

  // zero based index for the target and background indivudals

  map<int, int> it, ib;

  // deltaaf is the difference of allele frequency we bother to look at

  string deltaaf ;
  double daf  = 0;

  //

  int counts = 0;

  // type pooled GL PL

  string type = "NA";

    const struct option longopts[] =
      {
	{"version"   , 0, 0, 'v'},
	{"help"      , 0, 0, 'h'},
	{"cpu"       , 1, 0, 'x'},
	{"per"       , 1, 0, 'n'},
	{"file"      , 1, 0, 'f'},
	{"target"    , 1, 0, 't'},
	{"background", 1, 0, 'b'},
	{"region"    , 1, 0, 'r'},
	{"type"      , 1, 0, 'y'},

	{0,0,0,0}
      };

    int index;
    int iarg=0;

    while(iarg != -1)
      {
	iarg = getopt_long(argc, argv, "n:r:d:t:b:f:y:x:hv", longopts, &index);
	switch (iarg)
	  {
	  case 'h':
	    {
	      printHelp();
	      return 0;
	    }
	  case 'n':
	    {
	      nper = atoi(optarg);
	      cerr << "INFO: pVst will do " << nper << "permutations." << endl;
	      break;
	    }
	  case 'v':
	    {
	      printVersion();
	      return 0;
	    }
	  case 'y':
	    {
	      type = optarg;
	      cerr << "INFO: Copy number will be found in : " << type << endl;
	      break;
	    }
	  case 'x':
	    {
	      cerr << "INFO: OpenMp will use " << optarg << " cpus. " << endl;
	      cpu = atoi(optarg);
	      break;
	    }
	  case 't':
	    {
	      loadIndices(ib, optarg);
	      cerr << "INFO: There are " << ib.size() << " individuals in the target" << endl;
	      cerr << "INFO: target ids: " << optarg << endl;
	      break;
	    }
	  case 'b':
	    {
	      loadIndices(it, optarg);
	      cerr << "INFO: There are " << it.size() << " individuals in the background" << endl;
	      cerr << "INFO: background ids: " << optarg << endl;
	      break;
	    }
	  case 'f':
	    {
	      cerr << "INFO: File: " << optarg  <<  endl;
	      filename = optarg;
	      break;
	    }
	  case 'd':
	    {
	      cerr << "INFO: only scoring sites where the allele frequency difference is greater than: " << optarg << endl;
	      deltaaf = optarg;
	      daf = atof(deltaaf.c_str());
	      break;
	    }
	  case 'r':
            cerr << "INFO: set seqid region to : " << optarg << endl;
	    region = optarg;
	    break;
	  default:
	    break;
	  }
      }

#if defined HAS_OPENMP
    omp_set_num_threads(cpu);
#endif


    if(filename == "NA"){
      cerr << "FATAL: did not specify the file\n";
      printHelp();
      exit(1);
    }

    variantFile.open(filename);

    if(region != "NA"){
      if(! variantFile.setRegion(region)){
	cerr <<"FATAL: unable to set region" << endl;
	return 1;
      }
    }

    if (!variantFile.is_open()) {
      exit(1);
    }

    if(type == "NA"){
      cerr << "FATAL: failed to specify copy number genotype field" << endl;
      printHelp();
      return 1;
    }

    Variant var(variantFile);

    vector<string> samples = variantFile.sampleNames;
    int nsamples = samples.size();

    std::vector<copyNcounts*> dataBin;

    while (variantFile.getNextVariant(var)) {

      if(var.alt.size() > 1){
	continue;
      }

      std::map<string, bool> formatMap;
      for(const auto& itz : var.format){
		formatMap[itz] = true;
      }

      if(formatMap.find(type) == formatMap.end()){
	continue;
      }


      copyNcounts * varDat = new copyNcounts;

      varDat->pos   = var.position    ;
      varDat->seqid = var.sequenceName;
      varDat->type  = var.alt.front() ;

      if(var.info.find("CALLERS") != var.info.end()){
	stringstream caller;
	for(std::vector<std::string>::iterator z = var.info["CALLERS"].begin();
	    z != var.info["CALLERS"].end(); z++){
	  caller << (*z);
	  caller << ',';
	}

	varDat->callers = caller.str();
      }

      if(var.info.find("END") != var.info.end()){
	varDat->end = atol(var.info["END"].front().c_str());
      }
      else{
	varDat->end = varDat->pos;
      }

      vector < map< string, vector<string> > > target, background, total;

      int index = 0;


      for(int nsamp = 0; nsamp < nsamples; nsamp++){

	map<string, vector<string> > sample = var.samples[ samples[nsamp]];

	if(sample[type].front() != "."){
	  if(it.find(index) != it.end() ){
	    target.push_back(sample);
	    total.push_back(sample);
	  }
	  if(ib.find(index) != ib.end()){
	    background.push_back(sample);
	    total.push_back(sample);
	  }
	}

	index += 1;
      }

      loadDat(varDat, type, target, background);

      if(varDat->target.size() < 2
	 || varDat->background.size() < 2){
	delete varDat;
	continue;
      }

      dataBin.push_back(varDat);

      // this odd pattern is for open MP ... later
      if(dataBin.size() == cpu){

#if defined HAS_OPENMP
#pragma omp parallel for schedule(dynamic, 1)
#endif
	for(int i = 0 ; i < dataBin.size(); i++){
	  calc(dataBin[i]);
	}
	for(int i = 0 ; i < dataBin.size(); i++){
	  std::cout << dataBin[i]->results.str() << "\t" << dataBin[i]->targetV << "\t" << dataBin[i]->backgroundV ;

	  if(!dataBin[i]->callers.empty()){
	    std::cout << "\t" << dataBin[i]->callers << std::endl;
	  }
	  else{
	    std::cout << std::endl;
	  }


	  delete dataBin[i];
	}
	dataBin.clear();
      }
    }
    return 0;
}
